core/src/main/java/org/elasticsearch/search/fetch/StoredFieldsContext.java

```diff
@@ -0,0 +1,181 @@
+
+    /**
+     * Gets the field names to load and return as part of the search request.
+     */
+    public List<String> fieldNames() {
+        return fieldNames;
+    }
+
+    /**
+     * Adds the field names {@code fieldNames} to the list of fields to load.
+     */
+    public StoredFieldsContext addFieldNames(List<String> fieldNames) {
+        if (fetchFields == false) {
+            throw new IllegalArgumentException("stored fields are disabled");
+        }
+        this.fieldNames.addAll(fieldNames);
+        return this;
+    }
+
+    /**
+     * Adds a field name {@code field} to the list of fields to load.
+     */
+    public StoredFieldsContext addFieldName(String field) {
+        if (fetchFields == false) {
+            throw new IllegalArgumentException("stored fields are disabled");
+        }
+        this.fieldNames.add(field);
+        return this;
+    }
+
+    /**
+     * Returns true if the stored fields should be fetched, false otherwise.
+     */
+    public boolean fetchFields() {
+        return fetchFields;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        StoredFieldsContext that = (StoredFieldsContext) o;
+
+        if (fetchFields != that.fetchFields) return false;
+        return fieldNames != null ? fieldNames.equals(that.fieldNames) : that.fieldNames == null;
+
+    }
+
+    @Override
+    public int hashCode() {
+        int result = fieldNames != null ? fieldNames.hashCode() : 0;
+        result = 31 * result + (fetchFields ? 1 : 0);
+        return result;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        this.fetchFields = in.readBoolean();
+        if (fetchFields) {
+            this.fieldNames = (List<String>) in.readGenericValue();
+        } else {
+            this.fieldNames = null;
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(fetchFields);
+        if (fetchFields) {
+            out.writeGenericValue(fieldNames);
+        }
+    }
+
+    public void toXContent(String preferredName, XContentBuilder builder) throws IOException {
+        if (fetchFields == false) {
+            builder.field(preferredName, _NONE_);
+        }
+        if (fieldNames != null) {
+            if (fieldNames.size() == 1) {
+                builder.field(preferredName, fieldNames.get(0));
+            } else {
+                builder.startArray(preferredName);
+                for (String fieldName : fieldNames) {
+                    builder.value(fieldName);
+                }
+                builder.endArray();
+            }
+        }
+    }
+
+    public static StoredFieldsContext fromList(List<String> fieldNames) {
+        if (fieldNames.size() == 1 && _NONE_.equals(fieldNames.get(0))) {
+            return new StoredFieldsContext(false, null);
+        }
```

Comment:
should we also validate that _none_ is not in the list if it has multiple elements?